File: haskell/project_euler/28_1001Spiral.hs
import IO

sum_spiralDiagonals n = 
	sum_spiralDiagonals' n 1 [1,1,1,1] [2,4,6,8]
	where
	sum_spiralDiagonals' :: Int -> Int -> [Int] -> [Int] -> Int
	sum_spiralDiagonals' n sumDiags diags next_diags = 
		if (diags !! 3) == n
			then sumDiags
			else
				sum_spiralDiagonals' n curr_sum new_diags new_next_diags
				where
				new_diags = zipWith (+) diags next_diags
				curr_sum = sumDiags + sum (new_diags)
				new_next_diags = map (+8) next_diags

main = do
	print $ sum_spiralDiagonals(1001*1001)
	{- 669171001 -}

File: python/project_euler/28_1001Spiral.py
#!/usr/bin/python

def spiralDiagonals (n):
	sumdiags = 1
	diags = [1,1,1,1]
	next_diags = [2, 4, 6, 8]

	while diags[3] != n:
		for i in [0,1,2,3]:
			diags[i] += next_diags[i]
			next_diags[i] += 8
		sumdiags += sum(diags)
	print sumdiags

spiralDiagonals (1001*1001)
# 669171001
